docs/source/jit_language_reference_v2.rst

```diff
@@ -0,0 +1,320 @@
+    `integer <https://docs.python.org/3/reference/lexical_analysis.html#integer-literals>`
+
+Evaluation of a literal yields an object of the appropriate type with the specific value
+(with approximations applied as necessary for floats). Literals are immutable, and multiple evaluations
+of identical literals may obtain the same object or distinct objects with the same value.
+
+Parenthesized Forms
+"""""""""""""""""""
+
+::
+
+    parenth_form ::=  '(' [expression_list] ')'
+
+A parenthesized expression list yields whatever the expression list yields. If the list contains at least one
+comma, it yields a ``Tuple``; otherwise, it yields the single expression inside the expression list. An empty
+pair of parentheses yields an empty ``Tuple`` object (``Tuple[]``).
+
+List and Dictionary Displays
+""""""""""""""""""""""""""""
+
+::
+
+    list_comprehension ::=  expression comp_for
+    comp_for           ::=  'for' target_list 'in' or_expr
+    list_display       ::=  '[' [expression_list | list_comprehension] ']'
+    dict_display       ::=  '{' [key_datum_list | dict_comprehension] '}'
+    key_datum_list     ::=  key_datum (',' key_datum)*
+    key_datum          ::=  expression ':' expression
+    dict_comprehension ::=  key_datum comp_for
+
+Lists and dicts can be constructed by either either listing the container contents explicitly or providing
+instructions on how to compute them via a set of looping instructions (i.e. a *comprehension*). A comprehension
+is semantically equivalent to using a for loop and appending to an ongoing list the expression of the comprehension.
+Comprehensions implicitly create their own scope to make sure the items of the target list do not leak into the
+enclosing scope. In the case that container items are explicitly listed, the expressions in the expression list
+are evaluated left-to-right. If a key is repeated in a ``dict_display`` that has a ``key_datum_list``, then, the
+resultant dictionary uses the value from the rightmost datum in the list that uses the repeated key.
+
+Primaries
+^^^^^^^^^
+
+::
+
+    primary ::=  atom | attributeref | subscription | slicing | call
+
+
+Attribute References
+""""""""""""""""""""
+
+::
+
+    attributeref ::=  primary '.' identifier
+
+
+``primary`` must evaluate to an object of a type that supports attribute references that has an attribute named
+``identifier``.
+
+Subscriptions
+"""""""""""""
+
+::
+
+    subscription ::=  primary '[' expression_list ']'
+
+
+The primary must evaluate to an object that supports subscription. If it is a ``List`` , ``Tuple``, or ``str``,
+the expression list must evaluate to an integer or slice. If it is a ``Dict``, the expression list must evaluate
+to an object of the same type as the key type of the ``Dict``. If the primary is a ``ModuleList``, the expression
+list must be an ``integer`` literal. If the primary is a ``ModuleDict``, the expression must be a ``stringliteral``.
```

Comment:
Do indices on ModuleList and ModuleDict still need to be literals?

Reply:
I would say yes; subscriptions can appear in other contexts in addition to the RHS of an assignment. Using non-literal subscripts is allowed for annotated assignment, but not for any subscription in which the primary is a `ModuleList` or `ModuleDict`.